applications/CompressiblePotentialFlowApplication/custom_processes/define_2d_wake_process.cpp

```diff
@@ -0,0 +1,299 @@
+    MarkKuttaElements();
+    // Mark the trailing edge element that is further downstream as wake
+    MartkWakeTrailingEdgeElement();
+}
+
+// This function sets the wake direction and normal for further computations
+const void Define2DWakeProcess::SetWakeDirectionAndNormal()
+{
+    // Reading the free_stream_velocity from the properties
+    const auto free_stream_velocity = mrBodyModelPart.GetProcessInfo().GetValue(FREE_STREAM_VELOCITY);
+    KRATOS_ERROR_IF(free_stream_velocity.size() != 3)
+        << "The free stream velocity should be a vector with 3 components!"
+        << std::endl;
+
+    // Computing the norm of the free_stream_velocity vector
+    const double norm = sqrt(inner_prod(free_stream_velocity, free_stream_velocity));
+
+    // The wake direction is the free stream direction
+    mWakeDirection = free_stream_velocity / norm;
+
+    // Rotating 90° to obtain the wake normal
+    mWakeNormal(0) = -mWakeDirection(1);
+    mWakeNormal(1) = mWakeDirection(0);
+    mWakeNormal(2) = 0.0;
+}
+
+// This function finds and saves the trailing edge for further computations
+const void Define2DWakeProcess::SaveTrailingEdgeNode()
+{
+    double max_x_coordinate = -1e30;
+    NodeIteratorType trailing_edge_node;
+    for (auto it_node = mrBodyModelPart.NodesBegin();
+         it_node != mrBodyModelPart.NodesEnd(); ++it_node) {
+        if (it_node->X() > max_x_coordinate) {
+            max_x_coordinate = it_node->X();
+            trailing_edge_node = it_node;
+        }
+    }
+    trailing_edge_node->SetValue(TRAILING_EDGE, true);
+    mTrailingEdgeNode = trailing_edge_node;
+}
+
+// This function checks which elements are cut by the wake and marks them as
+// wake elements
+const void Define2DWakeProcess::MarkWakeElements()
+{
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+    bool potentially_wake = false;
+    bool is_wake_element = false;
+    BoundedVector<double, 3> nodal_distances_to_wake = ZeroVector(3);
+
+    #pragma omp parallel for private(potentially_wake, is_wake_element, nodal_distances_to_wake)
+    for (int i = 0; i < static_cast<int>(root_model_part.Elements().size()); i++) {
+        ModelPart::ElementIterator it_elem = root_model_part.ElementsBegin() + i;
+
+        // Check if the element is touching the trailing edge
+        CheckIfTrailingEdgeElement(it_elem);
+
+        // Elements downstream the trailing edge can be wake elements
+        potentially_wake = CheckIfPotentiallyWakeElement(it_elem);
+
+        if (potentially_wake) {
+            // Compute the nodal distances of the element to the wake
+            nodal_distances_to_wake = ComputeNodalDistancesToWake(it_elem);
+
+            // Selecting the cut (wake) elements
+            is_wake_element = CheckIfWakeElement(nodal_distances_to_wake);
+
+            // Mark wake element and save their nodal distances to the wake
+            if (is_wake_element) {
+                #pragma omp critical
```

Comment:
I think that you don't need the critical in here. It should be enough to lock the node you are accessing to by doing: 
```
it_elem->GetGeometry()[i].SetLock()
it_elem->GetGeometry()[i].FastGetSolutionStepValue(DISTANCE) = nodal_distances_to_wake(i);
it_elem->GetGeometry()[i].UnSetLock()
```
Besides, I'd create a reference to the geometry to avoid the repeated `it_elem->GetGeometry()` access.